### PR TITLE
Improve Canberra metric for DDP

### LINKED ISSUE #1313: Improve Canberra metric for DDP
## 🚀 Feature

Current implementation of Canberra metric is not compatible with ddp. The idea is to improve this metric to work in parallel.

### PR DESCRIPTION
**Pull request by sdesrozis**
Fixes #1313 

Description:

Improvement of Canberra metric to work with ddp.

Check list:
* [x] New tests are added (if a new feature is added)
* [ ] New doc strings: description and/or example code are in RST format
* [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/contrib/metrics/regression/canberra_metric.py [hunk 1]
```diff
@@ -22,13 +25,21 @@ class CanberraMetric(_BaseRegression):
...
+    def __init__(
+        self, output_transform: Callable = lambda x: x, device: Union[str, torch.device] = torch.device("cpu")
+    ):
+        self._sum_of_errors = None
+        super(CanberraMetric, self).__init__(output_transform, device)
+
+    @reinit__is_reduced
     def reset(self):
-        self._sum_of_errors = 0.0
+        self._sum_of_errors = 0
```
Comment (vfdev-5) [inline ignite/contrib/metrics/regression/canberra_metric.py:36]: Just remarked that this metric does not use torch tensor for accumulators. Can you update that too ? Take a look at Accuracy for an example

File: ignite/contrib/metrics/regression/canberra_metric.py [hunk 2]
```diff
@@ -22,13 +25,21 @@ class CanberraMetric(_BaseRegression):
...
 
     def _update(self, output):
         y_pred, y = output
```
Comment (vfdev-5) [inline ignite/contrib/metrics/regression/canberra_metric.py:39]: We have to detach the output too as it is done for core metrics

File: ignite/contrib/metrics/regression/canberra_metric.py (hunk 1 above)
Comment (sdesrozis) [reply ignite/contrib/metrics/regression/canberra_metric.py]: @vfdev-5 done

File: ignite/contrib/metrics/regression/canberra_metric.py (hunk 2 above)
Comment (sdesrozis) [reply ignite/contrib/metrics/regression/canberra_metric.py]: @vfdev-5 done

File: ignite/contrib/metrics/regression/canberra_metric.py [hunk 3]
```diff
@@ -22,13 +25,21 @@ class CanberraMetric(_BaseRegression):
...
+    @reinit__is_reduced
     def reset(self):
-        self._sum_of_errors = 0.0
+        self._sum_of_errors = torch.tensor(0.0, device=self._device)
 
     def _update(self, output):
-        y_pred, y = output
+        y_pred, y = output[0].detach(), output[1].detach()
         errors = torch.abs(y.view_as(y_pred) - y_pred) / (torch.abs(y_pred) + torch.abs(y.view_as(y_pred)))
         self._sum_of_errors += torch.sum(errors).item()
```
Comment (vfdev-5) [inline ignite/contrib/metrics/regression/canberra_metric.py:41]: No need to call `.item()` if we would like to make it work on GPU

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the update !

File: ignite/contrib/metrics/regression/canberra_metric.py (hunk 3 above)
Comment (sdesrozis) [reply ignite/contrib/metrics/regression/canberra_metric.py]: @vfdev-5 done

**Review by vfdev-5 — APPROVED** [review]
LGTM! Thanks @sdesrozis !

**Comment by vfdev-5** [comment]
@sdesrozis TPU is failing. Seems like in the update step you miss a cast to metric device and can mix devices... See how it was done for other core metrics.

File: ignite/contrib/metrics/regression/canberra_metric.py [hunk 4]
```diff
@@ -22,13 +25,21 @@ class CanberraMetric(_BaseRegression):
...
+        super(CanberraMetric, self).__init__(output_transform, device)
+
+    @reinit__is_reduced
     def reset(self):
-        self._sum_of_errors = 0.0
+        self._sum_of_errors = torch.tensor(0.0, device=self._device)
 
     def _update(self, output):
         y_pred, y = output
         errors = torch.abs(y.view_as(y_pred) - y_pred) / (torch.abs(y_pred) + torch.abs(y.view_as(y_pred)))
```
Comment (vfdev-5) [inline ignite/contrib/metrics/regression/canberra_metric.py:40]: I'm wondering now why we used `view_as` ... Docstring states that both should be of the same shape.

File: ignite/contrib/metrics/regression/_base.py [hunk 5]
```diff
@@ -38,13 +38,13 @@ class _BaseRegression(Metric):
     def update(self, output):
         _check_output_shapes(output)
         _check_output_types(output)
-        y_pred, y = output
+        y_pred, y = output[0].detach(), output[1].detach()
 
         if y_pred.ndimension() == 2 and y_pred.shape[1] == 1:
-            y_pred = y_pred.squeeze(dim=-1)
+            y_pred = y_pred.squeeze(dim=-1).to(self._device)
```
Comment (vfdev-5) [inline ignite/contrib/metrics/regression/_base.py:44]: I'm not sure about that. We can pass to else and thus skip device set. Let's keep it as for other metrics...

File: ignite/contrib/metrics/regression/canberra_metric.py (hunk 4 above)
Comment (sdesrozis) [reply ignite/contrib/metrics/regression/canberra_metric.py:40]: That is the current implementation. I can try to remove this.

Comment (vfdev-5) [reply ignite/contrib/metrics/regression/canberra_metric.py:40]: Yes, I know that it was here before. I'm just wondering now what was the good reason to do that.

Comment (sdesrozis) [reply ignite/contrib/metrics/regression/canberra_metric.py]: @vfdev-5 done, let's see.

File: ignite/contrib/metrics/regression/_base.py (hunk 5 above)
Comment (sdesrozis) [reply ignite/contrib/metrics/regression/_base.py]: @vfdev-5 done

File: ignite/contrib/metrics/regression/canberra_metric.py (hunk 4 above)
Comment (vfdev-5) [reply ignite/contrib/metrics/regression/canberra_metric.py]: Seems like we used that `view_as` everywhere for other regression metrics. We have to provide a uniform approach in other PRs.

File: ignite/metrics/__init__.py [hunk 6]
```diff
@@ -8,7 +8,15 @@
...
 from ignite.metrics.mean_squared_error import MeanSquaredError
-from ignite.metrics.metric import BatchFiltered, BatchWise, EpochWise, Metric, MetricUsage
+from ignite.metrics.metric import (
+    BatchFiltered,
+    BatchWise,
+    EpochWise,
+    Metric,
+    MetricUsage,
+    reinit__is_reduced,
+    sync_all_reduce,
```
Comment (vfdev-5) [inline ignite/metrics/__init__.py:18]: If you expose these methods as `ignite.metrics.reinit__is_reduced/sync_all_reduce`, can you also check locally the docs and the links on them here : https://pytorch.org/ignite/metrics.html#ignite.metrics.metric.sync_all_reduce
and maybe add links to those methods here : https://pytorch.org/ignite/metrics.html#metrics-and-distributed-computations
currently they are not clickable...

Comment (sdesrozis) [reply ignite/metrics/__init__.py:18]: I did it due to pycharm but maybe it is not necessary to expose.

Comment (vfdev-5) [reply ignite/metrics/__init__.py:18]: OK, let's keep them where they are. And I create an issue to fix unclickable link on them.

Comment (vfdev-5) [reply ignite/metrics/__init__.py:18]: I mean, let's not expose those methods

**Review by vfdev-5 — APPROVED** [review]
LGTM! Thanks @sdesrozis !

**Comment by vfdev-5** [comment]
@sdesrozis could you please investigate why there is a warning here : https://travis-ci.org/github/pytorch/ignite/jobs/730492404#L2924 
thanks !
